File: 자료구조/Python/백준-우선순위큐-1135.py
from sys import stdin

tree = [[] for _ in range(51)]

def solve(node):
    if not tree[node]:
        return 1

    child = []
    for i in range(len(tree[node])):
        child.append(solve(tree[node][i]) + 1)

    child.sort(reverse=True)

    ret = 0
    for i in range(len(child)):
        ret = max(ret, child[i] + i)

    return ret

def main():
    N = int(stdin.readline())
    numbers = list(map(int, stdin.readline().split()))

    for i in range(1, len(numbers)):
        tree[numbers[i]].append(i)

    print(solve(0) - 1)


if __name__ == "__main__":
    main()

File: 자료구조/C++/백준-우선순위큐-1135.cpp
# include <iostream>
# include <vector>
# include <algorithm>

using namespace std;

int n;
vector<int> tree[51];

int solve(int node) 
{
    if (tree[node].empty()) 
    {
        return 1;
    }
    
    vector<int> ch;

    for (int i = 0; i < tree[node].size(); i++) 
    {
        ch.push_back(solve(tree[node][i]) + 1);
    }

    sort(ch.begin(), ch.end(), greater<int>());

    int ret = 0;
    for (int i = 0; i < ch.size(); i++) { ret = max(ret, ch[i] + i); }

    return ret;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N, num;
    cin >> N;
    
    for (int i = 0; i < N; i++) 
    {
        cin >> num;
        if (i != 0) { tree[num].push_back(i); }
    }

    cout << solve(0) - 1 << endl;

    return 0;
}